7-gce/gce/deployment_manager/bookshelf.jinja

```diff
@@ -0,0 +1,156 @@
+{% set SERVICE = "bookshelf-" + env["deployment"] + "-frontend" %}
+
+#
+# Instance group setup
+#
+
+# First we have to create an instance template.
+# This template will be used by the instance group
+# to create new instances.
+resources:
+- name : {{ NAME }}
+  type: compute.v1.instanceTemplate
+  properties:
+    properties:
+      tags:
+        items:
+          - http-server
+      disks:
+        - boot: True
+          type: PERSISTENT
+          initializeParams:
+            sourceImage: {{ properties['machine-image'] }}
+            diskSizeGb: 10
+            diskType: pd-ssd
+      machineType: {{ properties['machine-type'] }}
+      serviceAccounts:
+          - email: default
+            scopes: {{ properties['scopes'] }}
+      metadata:
+        items:
+          - key: startup-script
+            value: |
+{{imports['startup-script']|indent(14, true)}}
+      networkInterfaces:
+          - network: global/networks/default
+            accessConfigs:
+              - type: ONE_TO_ONE_NAT
+                name: External NAT
+
+# Creates the managed instance group. This is responsible for creating
+# new instances using the instance template, as well as providing a named
+# port the backend service can target
+- name: {{ NAME }}-frontend-group
+  type: compute.v1.instanceGroupManager
+  properties:
+    instanceTemplate: $(ref.{{  NAME  }}.selfLink)
+    baseInstanceName: frontend-group
+    targetSize: 3
+    zone: {{ properties['zone'] }}
+    namedPorts:
+      - name: http
+        port: 8080
+
+
+
+# Load Balancer Setup
+#
+
+# A complete HTTP load balancer is structured as follows:
+#
+# 1) A global forwarding rule directs incoming requests to a target HTTP proxy.
+# 2) The target HTTP proxy checks each request against a URL map to determine the
+#    appropriate backend service for the request.
+# 3) The backend service directs each request to an appropriate backend based on
+#    serving capacity, zone, and instance health of its attached backends. The
+#    health of each backend instance is verified using either a health check.
+#
+# We'll create these resources in reverse order:
+# service, health check, backend service, url map, proxy.
+
+# Create a health check
+# The load balancer will use this check to keep track of which instances to send traffic to.
+# Note that health checks will not cause the load balancer to shutdown any instances.
+- name: {{ NAME }}-health-check
+  type: compute.v1.httpHealthCheck
+  properties:
+    requestPath: /_ah/health
+    port: 8080
+
+# Create a backend service, associate it with the health check and instance group.
+# The backend service serves as a target for load balancing.
+- name: {{ SERVICE }}
+  type: compute.v1.backendService
+  properties:
+    healthChecks:
+      - $(ref.{{ NAME }}-health-check.selfLink)
+    portName: http
+    backends:
+      - group: $(ref.{{ NAME }}-frontend-group.instanceGroup)
+        zone: {{ properties['zone'] }}
```

Comment:
One of the big benefits of L7 load balancing on GCE is that you can load balance across multiple regions. In my template I create IGMs for each zone in a list of zones, which might be cool to demonstrate here.... Otherwise I would say just use L3 since the urlmap is trivial.